Mathlib/Data/Matrix/Rank.lean

```diff
@@ -18,17 +18,91 @@ This definition does not depend on the choice of basis, see `Matrix.rank_eq_finr
 ## Main declarations
 
 * `Matrix.rank`: the rank of a matrix
+* `Matrix.cRank`: the rank of a matrix as a cardinal
+* `Matrix.eRank`: the rank of a matrix as a term in `ℕ∞`.
 
 -/
 
-
 open Matrix
 
 namespace Matrix
 
-open Module
+open Module Cardinal Set Submodule
+
+variable {l m n o R : Type*}
+
+section Infinite
+
+variable {m₀ n₀ : Type*} [Semiring R] {A : Matrix m n R}
+
+/-- The rank of a matrix, defined as the dimension of its column space, as a cardinal. -/
+noncomputable def cRank (A : Matrix m n R) : Cardinal := Module.rank R <| span R <| range Aᵀ
+
+lemma cRank_toNat_eq_finrank (A : Matrix m n R) :
+    A.cRank.toNat = Module.finrank R (span R (range Aᵀ)) := rfl
+
+lemma cRank_mono_col (A : Matrix m n R) (c : n₀ → n) : (A.submatrix id c).cRank ≤ A.cRank := by
+  apply Submodule.rank_mono <| span_mono ?_
+  rintro _ ⟨x, rfl⟩
+  exact ⟨c x, rfl⟩
+
+lemma cRank_lift_mono_row.{u,u₀,v} {m : Type u} {m₀ : Type u₀} {R : Type v} [Semiring R]
+    (A : Matrix m n R) (r : m₀ → m) :
+    lift.{u, max u₀ v} (A.submatrix r id).cRank ≤ lift.{u₀, max u v} A.cRank := by
+  let f : (m → R) →ₗ[R] (m₀ → R) := (LinearMap.funLeft R R r)
+  have h_eq : Submodule.map f (span R (range Aᵀ)) = span R (range (A.submatrix r id)ᵀ) := by
+    rw [LinearMap.map_span, ← image_univ, image_image, transpose_submatrix]
+    aesop
+  rw [cRank, ← h_eq]
+  have hwin := lift_rank_map_le f (span R (range Aᵀ))
+  simp_rw [← lift_umax] at hwin ⊢
+  exact hwin
+
+lemma cRank_mono_row.{u} {m m₀ : Type u} (A : Matrix m n R) (r : m₀ → m) :
+    (A.submatrix r id).cRank ≤ A.cRank  := by
+  simpa using A.cRank_lift_mono_row r
```

Comment:
Is it possible to unify these, giving a `cRank_submatrix_le` where you look at `A.submatrix r c` at once? I see that this is done for `eRank` below, so I expect there is some obstruction, but I don't understand what it is...

Reply:
I think we should have:
*  `lift_cRank_submatrix_le` (with fully general universes)
*  `cRank_submatrix_le` (with universe monomorphism, which is enough for `id`)

Reply:
I did try this, and got frustrated by the universe issues in the full generality of the submatrix statement. I can have another go.